resolve.c

```diff
@@ -85,20 +113,20 @@ void resolveClients(bool onlynew)
 		if(onlynew && !clients[clientID].new)
 			continue;
 
-		// Lock data when obtaining IP of this client
-		lock_shm();
-		const char* ipaddr = getstr(clients[clientID].ippos);
-		unlock_shm();
+		// Obtain/update hostname of this client
+		size_t oldnamepos = clients[clientID].namepos;
+		size_t newnamepos = resolveAndAddHostname(clients[clientID].ippos, oldnamepos);
 
-		// Important: Don't hold a lock while resolving as the main thread
-		// (dnsmasq) needs to be operable during the call to resolveHostname()
-		const char* hostname = resolveHostname(ipaddr);
+		if(newnamepos != oldnamepos)
+		{
+			// Need lock when storing obtained hostname
+			lock_shm();
+			clients[clientID].namepos = newnamepos;
+			unlock_shm();
+		}
 
-		// Finally, lock data when storing obtained hostname
-		lock_shm();
-		clients[clientID].namepos = addstr(hostname);
+		// Mark entry as not new
 		clients[clientID].new = false;
```

Comment:
This should be wrapped in a lock (and the forwarded call below).

Sorry for the third review!